Fix burr() distribution stats() method and add other methods

rlucas7 opened the pull request:
closes scipy issue #9544 (mean and stats method of burr() return erroneous results).

Also fixes the issues in fisk and burr12 that inherit from burr.

Adds unit test(s) for c<1 case (mean/variance are np.nan and c>1 case (mean is finite).

Also cleans up the docstring for burr and adds refs.

Retry of PR #9584 after rebase fail. See that PR for some discussion and comments, mostly but not entirely addressed. 

https://github.com/scipy/scipy/pull/9584

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
 
+    def _logpdf(self, x, c, d):
+        return (np.log(c) + np.log(d) + (-c - 1.0) * np.log(x) +
+                 (-d - 1.0) * np.log1p(x**(-c)))
```
`xlogy(-c-1, x)`

rlucas7 replied:
good callout

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
+                 (-d - 1.0) * np.log1p(x**(-c)))
+
     def _cdf(self, x, c, d):
         return (1 + x**(-c))**(-d)
```
Maybe route this through logcdf

rlucas7 replied:
Yes, should improve numerical stability. 

will change `return (1 + x**(-c))**(-d)` to `return np.exp(self._logcdf(x,c,d))`

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
+        return -d * np.log1p(x**(-c))
+
+    def _sf(self, x, c, d):
+        return 1 - (1 + x**(-c))**(-d)
```
Maybe add _logsf using log1p, too?

rlucas7 replied:
yes. Will switch `return 1 - (1 + x**(-c))**(-d)` to `return np.exp(self._logsf(x,c,d))`

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -782,12 +788,41 @@ def _pdf(self, x, c, d):
+        mu2 = np.where(c > 2.0, mu2_if_c, np.nan)
+        g1 = _lazywhere(c > 3.0, (c, e1, e2, e3, mu2_if_c),
+             lambda c, e1, e2, e3, mu2_if_c:
+             (e3 - 3 * e2 * e1 + 2 * e1**3) / np.sqrt((mu2_if_c)**3),
```
Nit: I feel grouping operators by priority a bit more readable. Like so, `e3 - 3*e2*e1 + ...`. No need to change it here though.

rlucas7 replied:
These are the raw moments, e.g. expected value of `x`, `E(X) = e1` second raw moment of `x`, `E(X^2) = e2` etc.
Do I understand correctly that your recommendation is to do the calculations directly on the  mean, variance, skewness, and kurtosis? I think that would be more code but I could do it if needed, would a comment here defining `e1`, `e2` etc help readability? Something like
```
#ek is the kth raw moment, e.g. e1 is the mean e2-e1**2 is variance etc.  
```

ev-br replied:
Nope, my comment is way more trivial, and is only about whitespace.

ev-br commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3450,6 +3450,18 @@ def test_burr12_ppf_small_arg():
+def test_burr_stats():
+    c, d = 5.0, 3
+    mean, variance = stats.burr(c, d).stats()
+    mean_hc, variance_hc = 1.4110263183925857, 0.22879948026191643
```
An obligatory question about hardcoded numbers: how to reproduce them? IOW, if this test starts failing in the future, how does a future maintainer check these numbers?

rlucas7 replied:
added a comment, these should come from the formula for mean and variance: 

https://edoc.unibas.ch/61230/1/20180305141821_5a9d439d76be5.pdf

page 6 or 8 depending on if you use pdf viewer page numbering or the numbering written on each page of the document. Eqn (7) gives the Kth moment. In the `burr_gen` class `b=1` and `p` in the pdf corresponds to `d` here in the code. The `a` in the pdf corresponds to the `c` here in the code. I'll double check these calculations tomorrow evening.

rlucas7 replied:
For the `mean_hc` and `variance_hc` I get:
```
>>> sc.beta(3 + 1/5, 1. - 1/5) * 3 
1.4110263183925855
>>> sc.beta(3 + 2/5, 1. - 2/5) * 3  - (sc.beta(3 + 1/5, 1. - 1/5) * 3)**2
0.2287994802619171
```

note there might be some loss in precision in the last few digits displayed. 
Which corresponds to what I would expect based on formulas in the manuscript:

https://edoc.unibas.ch/61230/1/20180305141821_5a9d439d76be5.pdf

and agrees with the output of `_stats` e.g. 

```

```

ev-br commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -3450,6 +3450,18 @@ def test_burr12_ppf_small_arg():
+    c, d = 5.0, 3
+    mean, variance = stats.burr(c, d).stats()
+    mean_hc, variance_hc = 1.4110263183925857, 0.22879948026191643
+    assert_allclose(mean_hc, mean_hc)
```
Typo,  `mean, mean_hc`

rlucas7 replied:
Yes, o/w this is the most useless unit test ever :)

rlucas7 commented:
I addressed some of the comments and pushed the changes. The moment calculations/unit tests should be (mostly) ok now. I'm still getting test failures locally on pdf - logpdf roundtrip comparisons. 
I'll try to address that tomorrow evening. Maybe the PR that @pvanmulbregt opened a couple hours ago addresses the numeric errors I'm getting?

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -757,37 +757,74 @@ class burr_gen(rv_continuous):
     .. math::
 
-        f(x, c, d) = c d x^{-c-1} (1+x^{-c})^{-d-1}
+        f(x, c, d) = c d x^{-c - 1} / (1 + x^{-c})^{d - 1}
```
Should this be `... (1 + x^{-c})^{d + 1}`?  That would match the implementation.

pvanmulbregt commented on scipy/stats/_continuous_distns.py:
```diff
@@ -906,14 +943,45 @@ def _pdf(self, x, c):
+             fillvalue=np.nan)
+        return mu, mu2, g1, g2
+
+    def _stats(self, c):
```
There are two implementations of `_stats(self, c)`.  Which one should be kept?

pvanmulbregt commented on scipy/stats/_distr_params.py:
```diff
@@ -10,7 +10,7 @@
     ['betaprime', (5, 6)],
     ['bradford', (0.29891359763170633,)],
-    ['burr', (10.5, 4.3)],
+    ['burr', (10.0, 5.0)],
```
Why change the default parameters?  Do they need to be integers with an integer difference?

rlucas7 replied:
I did this to make it easier for me to verify via hand calculations, e.g. I can do `\Gamma(9)` but I have difficulty caculating `\Gamma(9.5)` by pen and paper. I can change them back if preferred.

pvanmulbregt commented:
@rlucas7 I believe that all of this PR may now be incorporated into gh-9983. (Exception: I left the default testing parameters unchanged.) Is there anything missing?

rlucas7 commented:
Thanks for the ping, I’ll take a closer look tonight. 

Sincerely, 

-Lucas Roberts 

> On Apr 4, 2019, at 8:03 AM, Paul van Mulbregt <notifications@github.com> wrote:
> 
> @rlucas7 I believe that all of this PR may now be incorporated into gh-9983. (Exception: I left the default testing parameters unchanged.) Is there anything missing?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

rlucas7 commented:
Closing in favor of PR #9983
